delivery/services/file_system_service.py

```diff
@@ -41,6 +43,52 @@ def find_runfolder_directories(self, base_path):
         """
         return self.list_directories(base_path)
 
+    @staticmethod
+    def list_files_recursively(base_path):
+        for root, dirs, files in os.walk(base_path):
+            yield from map(lambda f: os.path.join(root, f), files)
+
+    @staticmethod
+    def extract_samplesheet_data(samplesheet_file):
+
+        def _extract_samplesheet_data_section(filehandle):
+            return list(csv.DictReader(filehandle))
+
+        try:
+            with open(samplesheet_file, "r") as fh:
+                while not next(fh).startswith("[Data]"):
```

Comment:
I liked this construct. Very elegant compared to solutions that I think I have written myself previously using a state variable instead.